app/code/community/Bolt/Boltpay/Helper/GraphQLTrait.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/**
+ * Bolt magento plugin
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Bolt
+ * @package    Bolt_Boltpay
+ * @copyright  Copyright (c) 2020 Bolt Financial, Inc (https://www.bolt.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+
+/**
+ * Trait Bolt_Boltpay_Helper_GraphQLTrait
+ *
+ * Provides utility methods related to the Bolt GraphQL API:
+ *
+ */
+trait Bolt_Boltpay_Helper_GraphQLTrait
+{
+
+    use Bolt_Boltpay_Helper_ApiTrait;
+
+    /**
+     * Make GraphQL call to Bolt server
+     *
+     * @param $query
+     * @param $operation
+     * @param $variables
+     * @return mixed
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    private function makeGqlCall($query, $operation, $variables)
+    {
+        try {
+
+            $gqlRequest = array(
+                "operationName" => $operation,
+                "variables" => $variables,
+                "query" => $query
+            );
+
+            $requestData = json_encode($gqlRequest, JSON_UNESCAPED_SLASHES);
+
+            $apiURL = $url = $this->getApiUrl() . Boltpay_GraphQL_Constants::MERCHANT_API_GQL_ENDPOINT;
+
+            $apiKey = Mage::getStoreConfig('payment/boltpay/api_key');
+
+            $headerInfo = $this->constructRequestHeaders($requestData, $apiKey);
+
+            $this->addMetaData(array('BOLT API REQUEST' => array('header' => $headerInfo)));
+            $this->addMetaData(array('BOLT API REQUEST' => array('data' => $requestData)), true);
+
+            $response = (string)$this->getApiClient()->post($apiURL, $requestData, $headerInfo)->getBody();
+
+            return json_decode($response);
+        } catch (\Exception $e) {
+            $this->notifyException($e);
```

Comment:
One more problem. We use methods from bugsnag trait in this trait, although formally we couldn't.
It doesn't trigger any issue because we don't use this trait without bugsnag.

It's not a new one, we use bugsnag in the same way in ApiTrait.

Should we fix this or not? If we should, I don't know how to refactor.

@mccottry

Reply:
@vitaliyreznikov 
As you have observed, the construct is not new.  And it is also not a problem.  You correctly observed the same long-standing pattern in `ApiTrait`.  In our design for helpers, we only have one concrete helper class, `Bolt_Boltpay_Helper_Data`, Bugsnag traits as well as all other traits will be available for use utilizing the `$this` identifier, as opposed to `$this->boltHelper()` that is used in all other classes.  You will also see this in your IDE if you are using something like PHPStorm.

In general, patterns that you see such as this are production and unit tested and are extremely unlikely to exist in the code base for months to years if there were null pointer issues.  However, there are always exceptions, (no pun intended), so it's awesome that you have an eye for this and asked questions about it.

Your current use of `$this->notifyException($e);` and `$this->logException($e);` in this context of Helper Traits, is correct and will execute without fail.